### PR TITLE
Default to a large fixed heap on systems we expect to benefit from it.

### PR DESCRIPTION
**Pull request by gbtitus**
On the target platforms with the most capable networks (InfiniBand, for
example) the ofi comm layer cannot use the best-performing providers
(verbs, for example) unless there is a fixed heap.  But in the past the
comm layer hasn't created a fixed heap by default, meaning that users
had to force it to do so through the `CHPL_RT_MAX_HEAP_SIZE` environment
variable in order to get good performance.

Here, when a tentative provider search indicates that a fixed heap could
improve performance by enabling use of a better provider, create one by
default and make it large: 85% of the compute node's physical memory.
We have to use a tentative provider search because we usually need to
make this decision early, when initializing the memory layer, which is
done long before the bulk of comm layer initialization.  And we need
this so early because some memory layers including the default jemalloc
based one initialize quite differently with and without a fixed heap.
"Tentative" here means that the provider search uses just the base
libfabric capabilities that we require.  In particular we don't look at
anything related to MCM conformance like the full provider search that
comes later does.

Users can override this by setting `CHPL_RT_MAX_HEAP_SIZE=0` in the
environment.

As part of this, adjust the verbose output from comm=ofi to include
whether or not a fixed heap exists and if so, how big it is.

This resolves https://github.com/Cray/chapel-private/issues/1959.
This resolves https://github.com/Cray/chapel-private/issues/1928.

### CODE REVIEW

File: runtime/include/chpl-format.h [hunk 1]
```diff
@@ -28,10 +28,11 @@ extern "C" {
...
 
 //
-// snprintf() a size as nnn[KMG]
+// These snprintf() a size as nnn[KMG] into the given buffer, but
+// unlike snprintf() itself, they return the buffer.
 //
-int chpl_snprintf_KMG_z(char* buf, int bufSize, size_t val);
-int chpl_snprintf_KMG_f(char* buf, int bufSize, double val);
+char* chpl_snprintf_KMG_z(char* buf, int bufSize, size_t val);
+char* chpl_snprintf_KMG_f(char* buf, int bufSize, double val);
```
Comment (ronawho) [inline runtime/include/chpl-format.h:35]: What's the motivation for returning the buffer? Just to make this cleaner to use in printf instead of having to make a call before the printf?

File: runtime/src/comm/ofi/comm-ofi-internal.h [hunk 2]
```diff
@@ -69,6 +69,7 @@ extern "C" {
   m(AMO_UNORD,              "AMOs: unordered operations")               \
   m(ACK,                    "tx acknowledgements")                      \
   m(ORDER,                  "ops done only for ordering")               \
+  m(HEAP,                   "layer-provided fixed heap")                \
```
Comment (ronawho) [inline runtime/src/comm/ofi/comm-ofi-internal.h:72]: Why is this needed? I don't think we want to be tracking the allocation used for a fixed heap as part of the memory interface.

**Review by ronawho — APPROVED** [review]
I had some inline comments/questions, but overall looks good.

I think we should again hold off on merging until nightly testing is more quiet.

File: runtime/include/chpl-format.h (hunk 1 above)
Comment (gbtitus) [reply runtime/include/chpl-format.h:35]: Yep, that's all.  I also considered doing versions that would print into an internal TLS buffer and return that, so the caller wouldn't have to supply the buffer.  That would be even cleaner, especially for situations (in comm=ofi, not ugni) where we're using these for debug output.  (Avoids an if-test on the debug flag.)  But would have to doc that you could only have one of these "outstanding" at a time, thus for the usage in comm=ugni it wouldn't be useful.

File: runtime/src/comm/ofi/comm-ofi-internal.h (hunk 2 above)
Comment (gbtitus) [reply runtime/src/comm/ofi/comm-ofi-internal.h:72]: It's used for debug output related to deciding whether or not to make a fixed heap and if so how big and where it starts.
